library/src/main/java/com/novoda/downloadmanager/RoomDownloadsPersistence.java

```diff
@@ -139,4 +140,46 @@ public boolean update(DownloadBatchId downloadBatchId, boolean notificationSeen)
         database.roomBatchDao().update(roomBatch);
         return true;
     }
+
+    @Override
+    public void persistCompletedBatch(CompletedDownloadBatch completedDownloadBatch) {
+        DownloadBatchId downloadBatchId = completedDownloadBatch.downloadBatchId();
+        DownloadBatchTitle downloadBatchTitle = completedDownloadBatch.downloadBatchTitle();
+        DownloadBatchStatus.Status downloadBatchStatus = DownloadBatchStatus.Status.DOWNLOADED;
+        long downloadedDateTimeInMillis = completedDownloadBatch.downloadedDateTimeInMillis();
+
+        DownloadsBatchPersisted persistedBatch = new LiteDownloadsBatchPersisted(
+                downloadBatchTitle,
+                downloadBatchId,
+                downloadBatchStatus,
+                downloadedDateTimeInMillis,
+                NOTIFICATION_SEEN
+        );
+        persistBatch(persistedBatch);
+
+        for (CompletedDownloadBatch.CompletedDownloadFile completedDownloadFile : completedDownloadBatch.completedDownloadFiles()) {
+            String url = completedDownloadFile.originalNetworkAddress();
+
+            String rawDownloadFileId = rawFileIdFrom(completedDownloadBatch.downloadBatchTitle(), completedDownloadFile);
+            DownloadFileId downloadFileId = DownloadFileIdCreator.createFrom(rawDownloadFileId);
+
+            DownloadsFilePersisted persistedFile = new LiteDownloadsFilePersisted(
+                    downloadBatchId,
+                    downloadFileId,
+                    new LiteFilePath(completedDownloadFile.newFileLocation()),
+                    completedDownloadFile.fileSize().totalSize(),
+                    url,
+                    FilePersistenceType.INTERNAL
+            );
+            persistFile(persistedFile);
+        }
+    }
+
+    private String rawFileIdFrom(DownloadBatchTitle batch, CompletedDownloadBatch.CompletedDownloadFile completedDownloadFile) {
+        if (completedDownloadFile.fileId() == null || completedDownloadFile.fileId().isEmpty()) {
+            return batch.asString() + System.nanoTime();
```

Comment:
Has `nanoTime()` enough precision so that if two sequential batch files don't have a valid fileId, this will give them two distinct ones?

Reply:
I'm thinking that this will bite us in the future, when we'll have phones with fast enough CPUs, so that _a lot_ of operations can be done in the same nanosecond.
Worth checking this approach maybe?
https://stackoverflow.com/a/15493695/1184680